src/duplicate_impl.rs

```diff
@@ -0,0 +1,160 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::fmt;
+
+/// An iterator adapter to filter out duplicate elements.
+///
+/// See [`.duplicate_by()`](../trait.Itertools.html#method.duplicate) for more information.
+#[derive(Clone)]
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct DuplicateBy<I: Iterator, V, F> {
+    iter: I,
+    // Use a hashmap for the entry API
+    used: HashMap<V, bool>,
+    f: F,
+}
+
+impl<I, V, F> fmt::Debug for DuplicateBy<I, V, F>
+    where I: Iterator + fmt::Debug,
+          V: fmt::Debug + Hash + Eq,
+{
+    debug_fmt_fields!(DuplicateBy, iter, used);
+}
+
+/// Create a new `DuplicateBy` iterator.
+pub fn duplicate_by<I, V, F>(iter: I, f: F) -> DuplicateBy<I, V, F>
+    where V: Eq + Hash,
+          F: FnMut(&I::Item) -> V,
+          I: Iterator,
+{
+    DuplicateBy {
+        iter,
+        used: HashMap::new(),
+        f,
+    }
+}
+
+impl<I, V, F> Iterator for DuplicateBy<I, V, F>
+    where I: Iterator,
+          V: Eq + Hash,
+          F: FnMut(&I::Item) -> V
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(v) = self.iter.next() {
+            let key = (self.f)(&v);
+            match self.used.get_mut(&key) {
+                None => { self.used.insert(key, false); },
+                Some(true) => (),
+                Some(produced) => {
+                    *produced = true;
+                    return Some(v);
+                },
+            }
+        }
+        None
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (low, hi) = self.iter.size_hint();
+        ((low > 0 && self.used.is_empty()) as usize, hi)
+    }
```

Comment:
Hi! Is this really correct? Let's assume `iter` is from a `Vec` containing `[1,2,3,4,5,6]` (so there are no duplicates). Then `self.iter.size_hint` is `(6, 6)` and `used` is initially empty, meaning that you'd mistakenly get `(1, 6)` instead of `(0, 6)`, right?

Reply:
Hi! You're right, I thought the hints would be the same as the Unique iterator. I just pushed the correct logic now and added a quickcheck test